body/index.ts

```diff
@@ -0,0 +1,102 @@
+import { Readable } from "stream";
+
+export type GetBodyOpts = {
+  encoding: string;
+  limit?: number;
+  expectedLength?: number;
+};
+
+export function getBody(stream: Readable, opts: GetBodyOpts): Promise<string> {
+  return new Promise((resolve, reject) => {
+    getBodySync(stream, opts, (err?: Error, buffer?: string) => {
+      if (err) return reject(err);
+      if (buffer) return resolve(buffer);
+      reject(new Error("body reader is in an unexpected state"));
+    });
+  });
+}
+
+type StreamReadCallback = (err?: Error, data?: string) => void;
+export function getBodySync(
+  stream: Readable,
+  opts: GetBodyOpts,
+  callback: StreamReadCallback,
```

Comment:
Do we want to keep the callback or would making this promise-returning work better? Unfortunately, we can't make this async and then throw inside the event handlers because the domain on the stream will catch it and tear everything down.